Recording exceeded limits

mohammadrabetian opened the pull request:
Hi @jsocol . I've been busy doing all these for this issue #210 .

So I've made a proxy and two handlers, one for cache and another for database backend as you said "I can definitely also see wanting to know which users had historically exceeded the limits, not just current." 
You can make as many as handlers you want, if there's any!

This functionality is completely separated from the core to not produce any security flaws.

Tests for the new functionality has been added and all the tests pass.

An abstract doc has been added for the settings. I wasn't sure if there's more explaining needed on the doc as it's simple and straight forward, but if needed, let me know.

I also did some re-structuring to the tests module because the test file was getting a Mega file.

jsocol commented:
Hi, sorry for not responding sooner—I wanted to say this looks like a really good direction! I gave it a read and there are probably some nuances in implementation, like avoiding accidentally running migrations for users who don't want to opt into this behavior.

It'll probably take me a couple of more weeks to do a good, thorough dive, just to set expectations. But I did see this and will follow-up and pursue it.

One global request, if you're willing: this project has preferred single quotes, and small as it might seem I'd rather that be consistent (except in cases where the quoted string contains a single quote, e.g. `"'"` > `'\''`). I noticed some places that convention was changed, and the new code generally uses double quotes.

Andrew-Chen-Wang commented on ratelimit/record_handlers/database.py:
```diff
@@ -0,0 +1,52 @@
+        user_agent = get_client_user_agent(request)
+
+        try:
+            limit_record = ExceededLimitRecord.objects.get(
+                user_agent=user_agent,
+                ip_address=client_ip_address,
+                username=username,
+                path_info=path_info,
+            )
+            limit_record.access_attempt_failures += 1
+            limit_record.save()
```
This can be made more efficient with [F expressions](https://docs.djangoproject.com/en/dev/ref/models/expressions/#f-expressions) and only getting the ID so we can update the record via the ID but not have to get all the values from the record (without the `only`). 
```suggestion
            limit_record = ExceededLimitRecord.objects.get(
                user_agent=user_agent,
                ip_address=client_ip_address,
                username=username,
                path_info=path_info,
            ).only("id")
            limit_record.access_attempt_failures = F("access_attempt_failures") + 1
            limit_record.save()
```
The "only" should be checked in a test. Test case can look like:
```python
with self.assertNumQueries(2):
    limit_record = ExceededLimitRecord.objects.get(
        user_agent=user_agent,
        ip_address=client_ip_address,
        username=username,
        path_info=path_info,
    ).only("id")
    limit_record.access_attempt_failures = F("access_attempt_failures") + 1
    limit_record.save()
```

mohammadrabetian replied:
Good suggestion!

Andrew-Chen-Wang commented on ratelimit/record_handlers/helpers.py:
```diff
@@ -0,0 +1,123 @@
+from ratelimit.conf import settings
+
+
+def get_client_ip_address(request) -> str:
```
There's already a tool that does that so that we don't need to install an external package (i.e. ipware): https://github.com/jsocol/django-ratelimit/blob/651c73074e43e6e501871baad0f1c2eb5cbbbe13/django_ratelimit/core.py#L29-L59

Andrew-Chen-Wang commented on ratelimit/record_handlers/helpers.py:
```diff
@@ -0,0 +1,123 @@
+    return request.META.get("PATH_INFO", "<unknown>")[:255]
+
+
+def get_client_username(request) -> str:
+    if request.user.is_authenticated:
+        return request.user.username
+    return "Anonymous"
```
Not everyone has a "username" field. You might want to look into `django.contrib.auth.models` where they define the Username field inside a list (specifically AbstractUser or AbstractBaseUser). I also think this should instead be based on user ID/pk instead of username to be more efficient (string lookup vs integer lookup).

If you're going to stick with username, then try this (but I'd still recommend using the integer pk AutoField):
```suggestion
from django.contrib.auth.models import AnonymousUser

def get_client_username(request) -> str:
    if request.user.is_authenticated:
        return getattr(request.user, request.user.USERNAME_FIELD)
    return AnonymousUser
```

Andrew-Chen-Wang commented on ratelimit/record_handlers/helpers.py:
```diff
@@ -0,0 +1,123 @@
+    return cache_timeout
+
+
+def make_cache_key(key_components):
+
+    cache_key_components = "".join(value for value in key_components.values() if value)
+    cache_key_digest = md5(cache_key_components.encode()).hexdigest()
+    cache_key = "drl-{}".format(cache_key_digest)
+
+    return cache_key
```
isn't this in `django_ratelimit.core`?

Andrew-Chen-Wang reviewed (changes requested):
Just added some of my thoughts and performance recommendations (though I think this is pretty out of date now):

jsocol commented:
This probably needs a bit of rework with the conflicts that have cropped up from some other refactors.

One thing I want to avoid is baking too much optional behavior into the "core." I wonder if we could tuck some of this into something like `ratelimit.ext.recorder` with its own `AppConfig`, so that installation requires explicit opt-in like `INSTALLED_APPS += ['ratelimit.ext.recorder']` and we don't end up installing unused models or causing inadvertent migrations.

Andrew-Chen-Wang commented:
Yes definitely doable. I think all you gotta do is just move the dirs and it should be fine (ref the token_blacklist app from here: https://github.com/jazzband/django-rest-framework-simplejwt/tree/master/rest_framework_simplejwt where the rest of the dir is not affiliated with any app)

mohammadrabetian commented on ratelimit/migrations/0001_initial.py:
```diff
@@ -0,0 +1,30 @@
+            options={
+                'verbose_name': 'exceeded limit record',
+                'verbose_name_plural': 'exceeded limit records',
+            },
```
The convention is to not touch the migration files! As we have consistent code signature everywhere else.

mohammadrabetian commented on ratelimit/migrations/0001_initial.py:
```diff
@@ -0,0 +1,30 @@
+        migrations.CreateModel(
+            name='ExceededLimitRecord',
+            fields=[
+                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
```
> This probably needs a bit of rework with the conflicts that have cropped up from some other refactors.
> 
> One thing I want to avoid is baking too much optional behavior into the "core." I wonder if we could tuck some of this into something like `ratelimit.ext.recorder` with its own `AppConfig`, so that installation requires explicit opt-in like `INSTALLED_APPS += ['ratelimit.ext.recorder']` and we don't end up installing unused models or causing inadvertent migrations.

I have updated the PR and fixed the conflicts. Sadly it's hard for me to find some spare time to work on this more but, I will make some more adjustments. I appreciate any help 👍

mohammadrabetian commented:
> This probably needs a bit of rework with the conflicts that have cropped up from some other refactors.
> 
> One thing I want to avoid is baking too much optional behavior into the "core." I wonder if we could tuck some of this into something like `ratelimit.ext.recorder` with its own `AppConfig`, so that installation requires explicit opt-in like `INSTALLED_APPS += ['ratelimit.ext.recorder']` and we don't end up installing unused models or causing inadvertent migrations.

I have updated the PR and fixed the conflicts. Sadly it's hard for me to find some spare time to work on this more but, I will make some more adjustments. I appreciate any help 👍
